compiler-core/src/language_server/code_action.rs

```diff
@@ -276,3 +276,148 @@ impl<'a> RedundantTupleInCaseSubject<'a> {
+    pattern_variables: Vec<EcoString>,
+}
+
+impl<'ast> ast::visit::Visit<'ast> for LetAssertToCase<'_> {
+    fn visit_typed_assignment(&mut self, assignment: &'ast ast::TypedAssignment) {
+        let range = src_span_to_lsp_range(assignment.location, &self.line_numbers);
+
+        // Only offer the code action if the cursor is over the statement
+        if !overlaps(range, self.params.range) {
+            return;
+        }
+
+        // This pattern only applies to `let assert`
+        if !matches!(assignment.kind, AssignmentKind::Assert { .. }) {
+            return;
+        };
+
+        // Get the source code for the tested expression
+        let location = assignment.value.location();
+        let expr = self
+            .module
+            .code
+            .get(location.start as usize..location.end as usize)
+            .expect("Location must be valid");
+
+        // Get the source code for the pattern
+        let pattern_location = assignment.pattern.location();
+        let pattern = self
+            .module
+            .code
+            .get(pattern_location.start as usize..pattern_location.end as usize)
+            .expect("Location must be valid");
+
+        let indent = " ".repeat(range.start.character as usize);
+
+        // Figure out which variables are assigned in the pattern
+        self.traverse_pattern(&assignment.pattern);
+        let variables = std::mem::take(&mut self.pattern_variables);
+
+        let assigned = match variables.len() {
+            0 => "",
+            1 => variables.first().expect("Variables is length one"),
+            _ => &format!("#({})", variables.join(", ")),
+        };
+
+        let edit = TextEdit {
+            range,
+            new_text: format!(
+                "let {assigned} = case {expr} {{
+{indent}  {pattern} -> {assigned}
+{indent}  _ -> panic
+{indent}}}",
+                // "_" isn't a valid expression, so we just return Nil from the case expression
+                assigned = if assigned == "_" { "Nil" } else { assigned }
+            ),
+        };
+
+        let uri = &self.params.text_document.uri;
+
+        CodeActionBuilder::new("Convert to case")
+            .kind(CodeActionKind::REFACTOR)
+            .changes(uri.clone(), vec![edit])
+            .preferred(true)
+            .push_to(&mut self.actions);
+    }
+}
+
+impl<'a> LetAssertToCase<'a> {
+    pub fn new(module: &'a Module, params: &'a CodeActionParams) -> Self {
+        let line_numbers = LineNumbers::new(&module.code);
+        Self {
+            module,
+            params,
+            actions: Vec::new(),
+            line_numbers,
+            pattern_variables: Vec::new(),
+        }
+    }
+
+    pub fn code_actions(mut self) -> Vec<CodeAction> {
+        self.visit_typed_module(&self.module.ast);
+        self.actions
+    }
+
+    // Recursively traverses a pattern and finds which variables get bound
+    fn traverse_pattern(&mut self, pattern: &TypedPattern) {
```

Comment:
How about we make a trait for visiting patterns? Then we wouldn't need to duplicate this code each time

Reply:
Should I add the pattern visiting code to the ast visitor, or create a separate trait?